### PR TITLE
docs: add kind constraints to the spec

### PR DESCRIPTION
**Pull request by jlapacik**
### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

### CODE REVIEW

**Comment by jlapacik** [comment]
@nathanielc first let me just say that I am in favor of replacing the term kind constraint with type constraint.

Now it is definitely not necessary to specify the various type constraints that Flux supports in a language spec, nor is it necessary to specify the individual types that make up a constraint. However this document has ballooned into much more than just a language spec. For one it houses the documentation for the entire standard library. One cannot say that the `range` function accepts input that is `Timeable` without mentioning the constraint or the types that define the constraint. If this is the one document for specifying the behavior of all builtin Flux functions, then you have to specify which type constraints are supported and what they mean.

If this document were to be purely a language spec and nothing more, then type constraints would not need to be mentioned at all. All that would be required is to specify that Flux is a type inferred language supporting both parametric and ad-hoc polymorphism. The latter allows for overloading certain operators like `+` while maintaining full static type safety.

I am in favor of splitting this document up so that it becomes a true language spec, but until then I think we need to specify the various type constraints that Flux supports and make sure that functions that take constrained input are documented correctly.

File: docs/SPEC.md [hunk 1]
```diff
@@ -567,6 +567,64 @@ This is structural polymorphism, objects of differing types can be used as the s
 
 This form of polymorphism means that these checks are performed during type inference and not during runtime. Type errors are found and reported before runtime.
 
+#### Kind Constraints
```
[the remark this answers is not in the document]
Comment (jlapacik) [reply docs/SPEC.md:570]: I'm definitely in favor of using the term type constraint.

[the remark this answers is not in the document]
Comment (pierwill) [reply docs/SPEC.md:570]: Another possibility might be `Traits` (inspired by Rust).

**Comment by nathanielc** [comment]
> I am in favor of splitting this document up so that it becomes a true language spec, but until then I think we need to specify the various type constraints that Flux supports and make sure that functions that take constrained input are documented correctly.

This make sense to me.  Let's add the specific constraints to the SPEC.

File: docs/SPEC.md [hunk 2]
```diff
@@ -567,6 +567,64 @@ This is structural polymorphism, objects of differing types can be used as the s
...
+
+Kind constraints are a type system concept that allow for constraining polymorphic types.
```
[the remark this answers is not in the document]
Comment (jlapacik) [reply docs/SPEC.md]: I updated the section on polymorphism, describing the three different forms that flux supports - parametric, record, and ad hoc.

**Pull request merged**